src/tools/pkcs11-tool.c

```diff
@@ -5762,101 +5762,64 @@ show_key(CK_SESSION_HANDLE sess, CK_OBJECT_HANDLE obj)
-				}
-			}
-			/* With BIT STRING remove unused bits in last byte indicator */
-			if (body_len > 0 && bytes[0] == 0x03) {
-				body_len--;
-				body++;
-			}
-
-			if (key_type == CKK_EC && body_len > 0) {
-				/*
-				 * (We only support uncompressed for now)
-				 * Uncompressed EC_POINT is DER OCTET STRING
-				 * or DER BIT STRING "04||x||y"
-				 * So a "256" bit key has x and y of 32 bytes each
-				 * something like: "03 42 00 04|x|y" or  "04 41 04||x||y"
-				 * Do simple size calculation based on DER encoding
-				 */
-				ksize = (body_len - 1) * 4;
-
-			} else if (body_len > 0) {
-				/*
-				 * EDDSA and XEDDSA in PKCS11 and only one coordinate
-				 */
-				ksize = (body_len) * 8;
-				if (ksize == 256)
-					ksize--; /* as 25519 uses 255 as bits */
+			if (point_bytes && point_size) {
+				printf("  EC_POINT:   ");
+				for (n = 0; n < point_size; n++)
+					printf("%02x", point_bytes[n]);
+				printf("\n");
 			}
 
-			if (ksize)
-				printf("  EC_POINT %lu bits\n", ksize);
-			else
-				printf("  EC_POINT size unknown");
-
-			if (bytes && body) {
-				if ((CK_LONG)size > 0) { /* Will print the point here */
-					printf("  EC_POINT:   ");
-					for (n = 0; n < body_len; n++)
-						printf("%02x", body[n]);
-					printf("\n");
-				}
-			}
-			free(bytes);
-			bytes = NULL;
-			size = 0;
-			bytes = getEC_PARAMS(sess, obj, &size);
-			if (bytes) {
-				if ((CK_LONG)size > 0) {
-					struct sc_object_id oid;
-
-					printf("  EC_PARAMS:  ");
-					for (n = 0; n < size; n++)
-						printf("%02x", bytes[n]);
-
-					if (size > 2 && bytes[0] == 0x06) { // OID
-						sc_init_oid(&oid);
-						if (sc_asn1_decode_object_id(bytes + 2, size - 2, &oid) == SC_SUCCESS) {
-							printf(" (OID %i", oid.value[0]);
-							if (oid.value[0] >= 0)
-								for (n = 1; (n < SC_MAX_OBJECT_ID_OCTETS)
-										&& (oid.value[n] >= 0); n++)
-									printf(".%i", oid.value[n]);
-							printf(")");
-						}
-					} else if (size > 2 && bytes[0] == 0x13) { // Printable string
-						printf(" (PrintableString %.*s)", bytes[1], bytes+2);
+			if (params_bytes && params_size > 2) {
+				struct sc_object_id oid;
+
+				printf("  EC_PARAMS:  ");
+				for (n = 0; n < params_size; n++)
+					printf("%02x", params_bytes[n]);
+
+				if (curve_info)
+					printf(" (\"%s\" OID:\"%s\")", curve_info->name, curve_info->oid);
+				else  /* if unknown curve, type and print something */
+				if (params_bytes[0] == SC_ASN1_OBJECT) {
+					sc_init_oid(&oid);
+					if (sc_asn1_decode_object_id(params_bytes + 2, size - 2, &oid) == SC_SUCCESS) {
+						printf(" (OID %i", oid.value[0]);
+						if (oid.value[0] >= 0)
+							for (n = 1; (n < SC_MAX_OBJECT_ID_OCTETS)
+									&& (oid.value[n] >= 0); n++)
+								printf(".%i", oid.value[n]);
+						printf(")");
 					}
-					printf("\n");
-
+				} else if (params_size > 2 && params_bytes[0] == SC_ASN1_PRINTABLESTRING) { // Printable string
+					printf(" (PrintableString %.*s)", params_bytes[1], params_bytes+2);
```

Comment:
it should be checked if the lenght in params_bytes[1] exceeds the buffer. Note, that OpenSC's ASN.1 handlers may be helpful here...

Reply:
The EC_PARMS returned value was already printed in hex so no need to try and parse it.